plugins/inputs/ctrlx_datalayer_sse/ctrlx_datalayer_sse.go

```diff
@@ -0,0 +1,303 @@
+		} else if s.acc != nil {
+			s.wg.Add(1)
+			s.acc.AddError(fmt.Errorf("received 'error' event for node: %v", em.Instance))
+			s.wg.Done()
+		}
+
+	case "keepalive":
+		// Keepalive events are ignored for the moment
+		s.Log.Debug("Received keepalive event")
+
+	default:
+		// Received a yet unsupported event type
+		s.Log.Debug("Received unsupported event: %v", se.Event)
+	}
+}
+
+// Init is for setup, and validating config
+func (s *CtrlXDataLayerSSE) Init() error {
+	for i, sub := range s.Subscription {
+		err := choice.Check(sub.QueueBehaviour, queueBehaviours)
+		if err != nil {
+			s.Log.Infof("The right queue behaviour values are %v", queueBehaviours)
+			return fmt.Errorf("sse subscription %v: setting 'queue_behaviour' %v is invalid", i, sub.QueueBehaviour)
+		}
+		err = choice.Check(sub.ValueChange, valueChanges)
+		if err != nil {
+			s.Log.Infof("The right value change values are %v", valueChanges)
+			return fmt.Errorf("sse subscription %v: setting 'value_change' %v is invalid", i, sub.ValueChange)
+		}
+		if len(sub.Nodes) == 0 {
+			s.Log.Warn("A configured subscription has no nodes configured")
+		}
+	}
+
+	// Generate valid communication url based on configured hostname
+	u := url.URL{
+		Scheme: "https",
+		Host:   s.Hostname,
+	}
+	s.url = u.String()
+	_, err := url.Parse(s.url)
+	if err != nil {
+		return fmt.Errorf("invalid hostname: %v, a valid hostname could be 'localhost', '198.168.1.1', 'host.example.com' or '10.0.2.2:8443'", s.Hostname)
+	}
+
+	return nil
+}
+
+// Start input as service, retain the accumulator, establish the connection
+// The reconnect will start if the connection is not established
+func (s *CtrlXDataLayerSSE) Start(acc telegraf.Accumulator) error {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+
+	s.connection = &http.Client{
+		Timeout: time.Duration(s.Timeout),
+	}
+	s.connection.Transport = &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: s.InsecureSkipVerify}}
+
+	s.tokenManager = tokenManager{
+		url:        s.url,
+		username:   s.Username,
+		password:   s.Password,
+		connection: s.connection,
+		log:        s.Log,
+	}
+
+	loginFailed := false
+	t, err := s.tokenManager.requestAuthToken()
+	if err != nil {
+		s.Log.Errorf("Hostname: %v, username: %v log in failed: %v", s.Hostname, s.Username, err)
+		loginFailed = true
+	}
+
+	s.acc = acc
+
+	for k, l := range s.Subscription {
+		l.index = k
+
+		if len(l.Nodes) == 0 {
+			continue
+		}
+
+		if loginFailed {
+			s.reconnect(l)
+			continue
+		}
+
+		subURL, err := s.createSubscription(&l)
+		if err != nil {
+			s.Log.Errorf("Failed to create subscription, error %v", err)
+		}
+		if subURL == "" {
+			s.reconnect(l)
+			continue
+		}
+
+		client := newSseClient(subURL, l, t.String(), s.InsecureSkipVerify, s.Log)
+		if s.autoReconnect {
+			client.onDisconnect(s.reconnect)
+		}
+		err = client.requestData(s, l.Measurement)
+		if err != nil {
+			s.reconnect(l)
+		}
+	}
+	return nil
+}
```

Comment:
I'm not sure I do understand the intended flow here. My understanding is

1. Get authentication token via `s.tokenManager.requestAuthToken()`. As long as we do not have that token, everything else (create subscription, gather data, ...) will not work.
2. Create a subscription on the ctrlX server and retrieve a URL where we can gather data.
3. Create a client for the subscription URL retrieved in step 2.
4. Loop forever (or until cancellation) and request data.

If this is the case, you should rather do something like
```suggestion
	s.acc = acc

	go gatherLoop(ctx)
	
	return nil
}

func (s *CtrlXDataLayerSSE) gatherLoop(ctx context.Context) {
	var authToken token

	// TODO: check for context cancelation here
	for {
		var err error
		
		authToken, err = s.tokenManager.requestAuthToken()
		if err == nil {
			break
		}
		s.Log.Errorf("Authentication for %s@%s failed: %v", s.Username, s.Hostname, err)
	}

	for k, l := range s.Subscription {
		l.index = k

		if len(l.Nodes) == 0 {
			continue
		}

		go func() {
			// TODO: check for context cancellation
			for {
				if err := s.reconnect(); err != nil {
					return fmt.Errorf("connecting to subscription, failed %w", err)
				}

				if err := client.requestData(s, l.Measurement); err != nil {
					s.Log.Errorf("Requesting data failed: %v", err)
				}
		}
	}
	return nil
}
```

Here `reconnect()` should handle the creation of the subscription and client but not requesting the data. Furthermore, it should fail for non-recoverable errors (if any). Of course  the code above misses context-cancellation code, you should add it.

Reply:
We tried to improve on this. Please have a look.